rules/value.go

```diff
@@ -0,0 +1,333 @@
+	}
+	rv.watch = watch
+	rv.status = runtimeValueStarted
+	rv.Unlock()
+
+	select {
+	case <-watch.C():
+	case <-time.After(rv.opts.InitWatchTimeout()):
+		err = errInitWatchTimeout
+	}
+
+	if err == nil {
+		err = rv.updateFn(watch.Get())
+	}
+
+	// NB(xichen): we want to start watching updates even though
+	// we may fail to initialize the value temporarily (e.g., during
+	// a network partition) so the value will be updated when the
+	// error condition is resolved
+	rv.updateWg.Add(1)
+	go rv.watchUpdates(watch)
+
+	return err
+}
+
+func (rv *runtimeValue) Update(value kv.Value) error {
+	rv.RLock()
+	if rv.status == runtimeValueClosed {
+		rv.RUnlock()
+		return errRuntimeValueClosed
+	}
+	initialized := rv.initialized
+	rv.RUnlock()
+
+	if !initialized {
+		return rv.initUpdate(value)
+	}
+	return rv.cachedUpdate(value)
+}
+
+func (rv *runtimeValue) Stop() error {
+	rv.Lock()
+	if rv.status != runtimeValueStarted {
+		rv.Unlock()
+		return errRuntimeValueNotStarted
+	}
+	if rv.watch != nil {
+		rv.watch.Close()
+	}
+	rv.watch = nil
+	rv.status = runtimeValueStopping
+	rv.Unlock()
+
+	// NB(xichen): we need to release lock while waiting or
+	// otherwise the stopping goroutine and the updating goroutine
+	// may end up in a deadlock situation
+	rv.updateWg.Wait()
+
+	rv.Lock()
+	defer rv.Unlock()
+
+	// The value may be closed at this point, in which case
+	// we don't override its status
+	if rv.status == runtimeValueStopping {
+		rv.status = runtimeValueStopped
+	}
+	return nil
+}
+
+func (rv *runtimeValue) Close() error {
+	rv.Lock()
+	if rv.status == runtimeValueClosed {
+		rv.Unlock()
+		return errRuntimeValueClosed
+	}
+	if rv.watch != nil {
+		rv.watch.Close()
+	}
+	rv.watch = nil
+	rv.status = runtimeValueClosed
+	rv.Unlock()
+
+	close(rv.closedCh)
+	rv.updateWg.Wait()
+
+	return nil
+}
+
+func (rv *runtimeValue) watchUpdates(watch kv.ValueWatch) {
+	defer rv.updateWg.Done()
+
+	for {
+		select {
+		case _, ok := <-watch.C():
+			if !ok {
+				return
+			}
+			if err := rv.updateFn(watch.Get()); err != nil {
+				rv.log.Errorf("error updating rules: %v", err)
+			}
+		case <-rv.closedCh:
+			return
+		}
+	}
+}
+
+func (rv *runtimeValue) initUpdate(update kv.Value) error {
+	latest, err := rv.unmarshalFn(update)
+	if err != nil {
+		return err
+	}
+	newVersion := rv.versionFn(latest)
+
+	// If this is the first time we update the rules, we need
+	// to go back in versions until we have found a version
+	// that's already in effect (or have fetched all versions)
+	var (
+		now         = rv.nowFn()
+		currVersion = newVersion - 1
+		currCutover = rv.cutoverFn(latest)
+		multiErr    = xerrors.NewMultiError()
+		historical  []interface{}
+	)
+
+	for ; currVersion > defaultVersion && currCutover.After(now); currVersion-- {
+		values, err := rv.store.History(rv.key, currVersion, currVersion+1)
+		if err != nil {
+			err = fmt.Errorf("error getting history for version %d: %v", currVersion, err)
+			multiErr = multiErr.Add(err)
+			continue
```

Comment:
I think so, this is so that if one version is corrupted in kv, we can still load the previous versions without preventing the entire set of historical versions from being loaded.